File: src/main/java/com/dude/dms/ui/components/dialogs/DocTextDialog.java
package com.dude.dms.ui.components.dialogs;

import com.dude.dms.backend.data.docs.Doc;
import com.dude.dms.backend.data.docs.TextBlock;
import com.dude.dms.backend.service.TextBlockService;
import com.vaadin.flow.component.dialog.Dialog;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.dom.Element;
import com.vaadin.flow.dom.ElementFactory;

import java.util.List;

public class DocTextDialog extends Dialog {

    public DocTextDialog(Doc doc, TextBlockService textBlockService) {
        VerticalLayout verticalLayout = new VerticalLayout();
        verticalLayout.setSizeFull();
        Element container = ElementFactory.createDiv();
        container.getStyle().set("width", "100%").set("height", "100%").set("position", "relative").set("maxWidth", "100%").set("maxHeight", "100%");
        verticalLayout.getElement().appendChild(container);
        add(verticalLayout);

        boolean first = true;
        List<TextBlock> textBlocks = textBlockService.findByDoc(doc);
        for (TextBlock textBlock : textBlocks) {
            if (first) {
                first = false;
                setSize(textBlock.getPageWidth(), textBlock.getPagHeight());
            }
            Element div = ElementFactory.createDiv(textBlock.getText());
            div.getStyle()
                    .set("fontSize", textBlock.getFontSize() / 6.0F + "vmin")
                    .set("textAlign", "center")
                    .set("position", "absolute")
                    .set("top", textBlock.getY() + "%")
                    .set("left", textBlock.getX() + "%")
                    .set("width", textBlock.getWidth() + "%")
                    .set("height", textBlock.getHeight() + "%");
            container.appendChild(div);
        }
    }

    private void setSize(float pageWidth, float pagHeight) {
        setWidth("60vw");
        setHeight(((pagHeight / pageWidth) * 100.0F) * 0.6F + "vh");
    }
}

File: src/main/java/com/dude/dms/ui/components/dialogs/DocTextDialog.kt
package com.dude.dms.ui.components.dialogs

import com.dude.dms.backend.data.docs.Doc
import com.dude.dms.backend.service.TextBlockService
import com.vaadin.flow.component.dialog.Dialog
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.dom.ElementFactory

class DocTextDialog(doc: Doc, textBlockService: TextBlockService) : Dialog() {

    init {
        val container = ElementFactory.createDiv().apply {
            style["width"] = "100%"
            style["height"] = "100%"
            style["position"] = "relative"
            style["maxWidth"] = "100%"
            style["maxHeight"] = "100%"
        }
        val verticalLayout = VerticalLayout().apply {
            setSizeFull()
            element.appendChild(container)
        }
        add(verticalLayout)

        var first = true
        for (textBlock in textBlockService.findByDoc(doc)) {
            if (first) {
                first = false
                setSize(textBlock.pageWidth, textBlock.pagHeight)
            }
            val div = ElementFactory.createDiv(textBlock.text).apply {
                style["fontSize"] = "${(textBlock.fontSize / 6.0f)}vmin"
                style["textAlign"] = "center"
                style["position"] = "absolute"
                style["top"] = "${textBlock.y}%"
                style["left"] = "${textBlock.x}%"
                style["width"] = "${textBlock.width}%"
                style["height"] = "${textBlock.height}%"
            }
            container.appendChild(div)
        }
    }

    private fun setSize(pageWidth: Float, pagHeight: Float) {
        width = "60vw"
        height = "${(pagHeight / pageWidth * 100.0f * 0.6f)}vh"
    }
}